architecture-decisions/0002-indexing.md

```diff
@@ -0,0 +1,86 @@
+
+For newly added records (not a full pipeline run of all records), we expect to see changes within five minutes of persistence in Figgy, as our stakeholders often do patron requests by "Completing" a record in Figgy and then sending a resource to a patron. They shouldn't have to wait more than 5 minutes to do that.
+
+```mermaid
+flowchart LR
+    A[Figgy Postgres] -->|Hydrate| B[DPUL-C Figgy Record Cache]
+    B -->|Transform| C[DPUL-C Solr Record Cache]
+    C -->|Index| D[DPUL-C Solr]
+```
+
+### Hydration
+
+Hydration will copy records from Figgy and place them into a cache in DPUL-Collections. This pattern will allow us to do the following steps no matter the uptime or performance characteristics of our source repository.
+
+Every minute DPUL-C will poll Figgy's `orm_resources` table for newly updated records and copy them into a local postgres cache that has the following structure:
+
+| id   | data  | log_order | figgy_modified_date |
+|------|-------|-----------|---------------------|
+| UUID | JSONB | INT       | DATETIME            |
+
+We'll pull records as well as DeletionMarkers so we'll know and record when records have been deleted from Figgy.
+
+#### Performance Requirements
+
+1 Hour - 2 Days
+
+##### Performance Reasoning
+
+The faster we can do a full re-harvest, the faster we can pull in broad metadata changes from upstream (such as new Figgy or Bibdata data.) We want these kinds of tickets to have at most two days of delay.
+
+### Transformation
+
+Every minute Transformation will poll the records cached by the Hydration step, convert them to a Solr document, and store that solr document in a local postgres cache with the following structure:
+
+| id   | data  | log_order |
+|------|-------|-----------|
+| UUID | JSONB | INT       |
+
+#### Performance Requirements
+
+30 minutes - 2 hours
+
+##### Performance Reasoning
+
+We will need to do a re-transformation when we add new fields to the index, which we expect to do often. The faster we can do that, the more of those tickets we can do. With a two hour transformation stage we can do more than one such transformation a day, significantly improving our productivity.
+
+### Reindex
+
+Every minute Reindex will poll the records cached by the Transformation step and index them into Solr as a batch.
+
+#### Performance Requirements
+
+10 minutes - 1 hour
+
+##### Performance Reasoning
+
+We expect reindexing to need to happen often - either because of changing weights in Solr, migrating Solr machines, or testing new configurations. By tightening up this time as much as possible we can try many different weights in a day, supporting our vision of being able to create a joyful discovery experience. We believe this performance estimate is reasonable given that there won't be any transformation necessary - it will go as fast as Solr can accept documents.
```

Comment:
There's still more to think about in the indexing step. How do we ensure we're not crashing other applications by flooding the solr box/es? Do we need changes to the existing solr cloud setup? Is there any of this that should be part of this ADR? Or is it all a separate concern?

Reply:
> There's still more to think about in the indexing step. How do we ensure we're not crashing other applications by flooding the solr box/es? Do we need changes to the existing solr cloud setup? Is there any of this that should be part of this ADR? Or is it all a separate concern?

:+1: Is "the indexing system should implement a way to apply back-pressure to allow us to scale the amount of processing we're demanding of Solr, to prevent our indexing from adversely affecting other applications" total nonsense or does it make sense?

Reply:
it sounds good; I assume you have some idea of how to do that?

Reply:
Uh, a few years ago it was "Flow": https://hexdocs.pm/flow/Flow.html. I dunno what the world is like now.

Reply:
Or Broadway with a custom producer connected to ecto? https://github.com/dashbitco/broadway and something local like https://github.com/maartenvanvliet/off_broadway_ecto